bucketstatscollector.go

```diff
@@ -0,0 +1,296 @@
+			nil,
+		),
+
+		Depth: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, "depth"),
+			"Number of levels in B+ tree for a bucket.",
+			labels,
+			nil,
+		),
+
+		PhysicalBranchPagesAllocatedBytes: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, "physical_branch_pages_allocated_bytes"),
+			"Number of bytes allocated in physical branch pages for a bucket.",
+			labels,
+			nil,
+		),
+
+		PhysicalBranchPagesInUseBytes: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, "physical_branch_pages_in_use_bytes"),
+			"Number of bytes in use in physical branch pages for a bucket.",
+			labels,
+			nil,
+		),
+
+		PhysicalLeafPagesAllocatedBytes: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, "physical_leaf_pages_allocated_bytes"),
+			"Number of bytes allocated in physical leaf pages for a bucket.",
+			labels,
+			nil,
+		),
+
+		PhysicalLeafPagesInUseBytes: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, "physical_leaf_pages_in_use_bytes"),
+			"Number of bytes in use in physical leaf pages for a bucket.",
+			labels,
+			nil,
+		),
+
+		Buckets: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, "buckets"),
+			"Number of buckets within a bucket, including the top bucket.",
+			labels,
+			nil,
+		),
+
+		InlinedBuckets: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, "inlined_buckets"),
+			"Number of inlined buckets for a bucket.",
+			labels,
+			nil,
+		),
+
+		InlinedBucketsInUseBytes: prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, "inlined_buckets_in_use_bytes"),
+			"Number of bytes in use for inlined buckets.",
+			labels,
+			nil,
+		),
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *bucketStatsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ds := []*prometheus.Desc{
+		c.LogicalBranchPages,
+		c.PhysicalBranchOverflowPages,
+		c.LogicalLeafPages,
+		c.PhysicalLeafOverflowPages,
+		c.Keys,
+		c.Depth,
+		c.PhysicalBranchPagesAllocatedBytes,
+		c.PhysicalBranchPagesInUseBytes,
+		c.PhysicalLeafPagesAllocatedBytes,
+		c.PhysicalLeafPagesInUseBytes,
+		c.Buckets,
+		c.InlinedBuckets,
+		c.InlinedBucketsInUseBytes,
+	}
+
+	for _, d := range ds {
+		ch <- d
+	}
+}
+
+// A forEachBucketStatsFunc is a function which is repeatedly called for all
+// buckets in a Bolt database to collect bucket statistics.
+type forEachBucketStatsFunc func(bucket string, s bolt.BucketStats) error
+
+// forEachWithBoltDB begins a read-only bolt transaction and returns a forEach
+// function for a bucketStatsCollector.  The returned function is invoked
+// repeatedly for each bucket and its stats retrieved from the Bolt database
+// handle.
+func forEachWithBoltDB(db *bolt.DB) func(forEachBucketStatsFunc) error {
+	return func(iter forEachBucketStatsFunc) error {
+		return db.View(func(tx *bolt.Tx) error {
+			return tx.ForEach(func(name []byte, b *bolt.Bucket) error {
```

Comment:
Off the top of my head, if you want to handle nested buckets, you have to iterate over all KV pairs in a bucket and if the value is nil for a given key, it *could* be a nested bucket. You'd check that using the Bucket.Bucket(key []byte) function.

The bottom line is that it's not impossible to do, but if you have a big DB and you don't already know where the nested buckets are, it can take a long time to find them since you'd basically have to traverse all the KV pairs in the DB!